[디디] API 테스트/문서자동화 미션 제출합니다.

fucct opened the pull request:
안녕하세요 리뷰어님! 
이번 미션 진행하면서 궁금했던 점들을 정리해보았습니다. 답변주시면 감사하겠습니다.😄

- 인수테스트와 api 테스트 둘 다 예외사항에 대한 테스트를 추가해야 하는지, 아니면 하나만 정해서 추가하면 되는지 정확하지 않네요.. 만약 정해서 한다면 어느곳에 두는게 적절할지도 답변주시면 감사하겠습니다
- 인수테스트에서 Junit5의 다이나믹 테스트를 이용하면 유용하다는 걸 알게 되었는데요, 문제는 이전 테스트에서 사용한 변수를 다음 테스트에서 사용할 수 없다는 점이였습니다. 방법이 있는데 제가 찾지 못한걸까요? 아니면 방법이 없지만 감수하고 사용하는 걸까요..?😭
- Spring data Jdbc 관련 질문입니다. 제가 DDD에 대한 기본 지식이 너무 부족해서 어떻게 보면 사소한 질문일수도 있습니다. 또한 미션을 진행하면서 제가 떠오른 궁금증이라 미션과 관련이 깊지 않을 것 같아서 질문 드려도 될 지 모르겠네요..😭  첫번째 질문은 어떠한 엔티티에 진입 가능한 root가 여러개라면, 해당 엔티티는 어떠한 Aggregate에도 포함되지 않는 엔티티인가요? 예를 들면 Article을 통해 그 게시물에 달린 Comment에 접근이 가능하고, User를 통해 User가 작성한 Comment에도 접근 가능하다면, Comment는 독립적인 Aggregate가 되는걸까요..? 
- 두번째 질문은, 제 질문이 맞다면, 각각의 관계에 있어서 Ref Table이 반드시 존재해야 하나요? 예를 들면 Article 이 삭제될 때 Comment도 같이 삭제되길 원하더라도 독립된 Aggregate이므로 ref Table 을 두고, 로직에서 삭제를 하는게 맞나요? 이렇게되면 하나의 Transaction에서 처리가 불가능할 것 같아서 궁금증이 생겼습니다. 

 리뷰 감사드립니다😄

fucct committed:
feat: 실습1, 2 구현

- 문서화 실습 구현
- 인증 실습 구현

fucct committed:
refactor: MemberResponse 기본 생성자 추가

fucct committed:
feat: MethodArgumentResolver에서 password와 password-check가 일치하는 지 검증

fucct committed:
feat: interceptor 와 methodArgumentResolver 에서 발생하는 예외 관리

- 아직 미완성..

fucct committed:
feat: annotation 기반 method argument validate 추가

- 패스워드 일치
- 이메일 중복

fucct committed:
feat: create 예외에 대한 문서화 구현

fucct committed:
feat: get에 대한 서비스, 컨트롤러 테스트와 문서화 구현

fucct committed:
feat: update에 대한 서비스, 컨트롤러 테스트와 문서화 구현

fucct committed:
feat: delete에 대한 서비스, 컨트롤러 테스트와 문서화 구현

fucct committed:
feat: 회원 가입 프론트 연동 구현

fucct committed:
feat: 로그인, 회원 정보 보기 기능 구현

fucct committed:
feat: 업데이트 구현

fucct committed:
feat: 탈퇴 구현

fucct committed:
refactor, feat: 테스트 정상화 및, 문서화 구현

- 전체 테스트 정상화
- 문서화 추가 양식 구현 (필수값 여부, 입력값 foramt)
- 예외 사항 추가

fucct committed:
feat: IsAuth 어노테이션 추가 및 테스트와 문서 수정

fucct committed:
feat: 컨트롤러 테스트 추가

- 토큰이 없는 경우 정보조회 테스트 추가

fucct committed:
feat: 즐겨찾기 서비스 테스트 추가

fucct committed:
feat: 즐겨찾기 관련 도메인 테스트 추가

fucct committed:
feat: 즐겨찾기 관련 도메인 테스트 추가

- 즐겨찾기 삭제 테스트 및 구현

fucct committed:
feat: 즐겨 찾기 프론트 연동 구현

fucct committed:
refactor: 테스트 정상화

fucct committed:
refactor: 패키지 수정

- 패키지 구조 수정

fucct committed:
refactor: 예외 관계 수정

- GlobalExceptionHandler 수정
- exception 상속 관계 수정

fucct committed:
refactor: 테스트 정상화

- 인터셉터 path 추가
- exception handler 수정

fucct committed:
docs: README.md 수정

fucct committed:
feat: Member 관련 인수테스트 추가

- 예외 사항에 대한 인수테스트 추가

fucct committed:
feat: Member 관련 api테스트 추가

- 예외사항에 대한 문서화 완료

fucct committed:
refactor: documentation 메서드 중복 제거

fucct committed:
refactor: favorites 예외 테스트 추가

-문서화 구현 완료

phs1116 commented on src/main/java/wooteco/subway/domain/favorite/Favorite.java:
```diff
@@ -0,0 +1,57 @@
+        return targetId;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o)
+            return true;
+        if (o == null || getClass() != o.getClass())
+            return false;
+        Favorite favorite = (Favorite)o;
+        return Objects.equals(sourceId, favorite.sourceId) &&
+            Objects.equals(targetId, favorite.targetId);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(sourceId, targetId);
+    }
```
Favorite을 식별하는 식별자는 station에 대한 id가 아니라 Favorite의 ID 아닐까요? :)
같은 sourceId, targetId이더라도 사용자에 따라 다른 객체이니까요.

fucct replied:
바로 수정했습니다 😅

phs1116 commented on src/main/java/wooteco/subway/domain/member/Member.java:
```diff
@@ -1,14 +1,21 @@
     private String email;
     private String name;
     private String password;
+    private Set<Favorite> favorites = new LinkedHashSet<>();
```
즐겨찾기를 멤버에서 책임질 필요가 있을까? 라는 고민이 조금 드네요.
도메인 상으로도 둘이 memberId를 공유한다는 것 외에 컨텍스트가 떨어져 있는 경향도 있어서요. :)

Favorite에 대해서든 별도 바운디드 컨텍스트로 쪼개서 Favorite이 memberId를 알고있고, 이로 인하여 비즈니스를 풀어나갈수도 있지 않을까 싶어요. :) 한번 고민해보세요.

fucct replied:
좋은 피드백 감사합니다! 제가 정확하게 피드백을 이해했는지 확실하지 않아서 확인좀 해주시면 감사하겠습니다! 의미상(?)으로는 멤버가 즐겨찾기를 가지고 있는게 맞지만, Member의 크기가 너무 커지기 때문에 비즈니스적으로 의미를 분리하라는 말씀이신가요? 제가 어제 책을 읽다가 도메인 모델과 살짝 다르더라도 구현시에 적합한 방식으로 연관관계를 설정하는게 옳다 라는 내용을 읽었는데 혹시 이 의도가 주신 피드백과 일치하는  지 여쭤보고싶네요 😀

phs1116 commented on src/main/java/wooteco/subway/domain/member/Member.java:
```diff
@@ -54,4 +74,18 @@ public void update(String name, String password) {
+            .filter(favorite -> favorite.isSameId(id))
+            .findFirst()
+            .ifPresent(favorites::remove);
+    }
```
Member에서 Favorite을 들고간다면 일급 컬렉션으로 분리해나갈수도 있을 것 같네요. :)

fucct replied:
우선 Favorite을 독립적으로 분리했습니다! 일급 컬렉션을 통해 구현하는 방법도 연습해봐야겠네요!

phs1116 commented on src/main/java/wooteco/subway/web/controller/FavoriteController.java:
```diff
@@ -0,0 +1,42 @@
+        this.favoriteService = favoriteService;
+    }
+
+    @IsAuth(isAuth = Auth.AUTH)
```
어노테이션 기반으로 인터셉터를 체크하도록 하셨네요. 👍 
Auth enun는 굳이 필요 없어 보이는데 그냥 인증이 필요하다는 의미의 어노테이션 하나만 두고 사용해도 되지 않을까요?

fucct replied:
블로그를 보면서 따라하다 보니 이렇게 됐네요..😭 다른 종류의 인증이 추가될 경우를 고려하여 작성한 것 같은데 이번 미션에서는 어노테이션 유무만으로 체크가 가능하겠네요! 감사합니다!

phs1116 commented on src/main/java/wooteco/subway/config/WebMvcConfig.java:
```diff
@@ -1,37 +1,31 @@
-        registry.addInterceptor(sessionInterceptor).addPathPatterns("/me/session");
-        registry.addInterceptor(bearerAuthInterceptor).addPathPatterns("/me/bearer");
+        registry.addInterceptor(bearerAuthInterceptor)
+            .addPathPatterns("/me", "/me/detail", "/me/favorites", "/members", "/members/*",
+                "/favorites", "/favorites/*");
```
어노테이션 기반, 그리고 URL기반 두가지로 설정이 되어있는데,
관리 포인트가 두가지로 늘어 헷갈릴수도 있을 것 같아요.
한쪽으로 통일하는것도 괜찮을 것 같아요. :)

fucct replied:
이 부분까진 생각하지 못했네요! 매 경우 인터셉터를 타도록 수정했습니다!

phs1116 commented on src/main/java/wooteco/subway/web/controller/GlobalMemberExceptionHandler.java:
```diff
@@ -0,0 +1,54 @@
+        return ResponseEntity.badRequest().body(new ErrorResponse(e.getMessage()));
+    }
+
+    @ExceptionHandler(value = DataAccessException.class)
+    public ResponseEntity<ErrorResponse> handlerDataAccessException(DataAccessException e) {
+        return ResponseEntity.badRequest().body(new ErrorResponse(e.getMessage()));
+    }
+
+    @ExceptionHandler(value = RuntimeException.class)
+    public ResponseEntity<ErrorResponse> handleRuntimeException(RuntimeException e) {
+
+        return ResponseEntity.badRequest().body(new ErrorResponse(e.getMessage(), e.getCause().getMessage()));
+    }
```
시스템 예외에 대해서 e.getMessage()로 그대로 클라이언트에게 전달하는 것은
시스템 보안상 내부 오류를 사용자한테 노출되기 떄문에 그닥 좋은 방법은 아니에요.
별도 메시지를 지정해서 내려주는게 좋을 것 같아요.

그리고 이 경우에 대해서는 내부 시스템 에러이기 떄문에 400대보다는 500 status code로 내려주는게 더 적합할 것 같네요. :)

phs1116 replied:
그리고 RuntimeException외에 Exception을 잡아서 처리해주는게 더 안전할 것 같아요.
두가지 경우는 정말 시스템 예외이기 떄문에 에러로그를 익셉션 핸들러에서 찍어주는게 좋아요.
여기서 예외를 먹어 내려주기 때문에 로그상에서 알기 어렵거든요. :)

Sl4fj를 적용해서 에러로그를 찍어보아요. :)

fucct replied:
오 안그래도 예외에 대해 어떤 내용을 보내줘야 하는가에 대한 궁금증이 있었습니다! 감사합니다! 추가적인 질문이 있는데요, e.getMessage가 보안상 좋은 방법이 아니라는 것은 이번 제 미션에서의 RuntimeException이나 Exception처럼 예상치 못한(?) 시스템 에러에 대해서만 해당하는 말씀이신가요? 아니라면 다른 예외에 대해서도 e.getMessage() 방식으로 보내주는 특정 예외 메세지도 좋은방식이 아니라는 건가요?  또한 500에러에 대해서는 사용자에게 알수 없는 에러라는 정보만 제공하면 되는걸까요? 제가 제대로 이해한건지 확신이 서질 않네요 😭

phs1116 replied:
다른 예외들에 담겨있는 메시지가 클라이언트에게 노출되어도 되는 메시지라면 e.getMessage()로 내려줘도 무방합니다. :)
사용자에게는 시스템에서 내려주는 에러 메시지를 보여주지 않는다는게 핵심이라고 생각하시면 될 것 같아요.
500이 모두 알수없는 예외는 아니고, 클라이언트에게 표현해줄 내용이 있다면 별도로 메시지를 내려줄순 있겠죠.

phs1116 commented on src/main/java/wooteco/subway/web/prehandler/AuthorizationExtractor.java:
```diff
@@ -0,0 +1,26 @@
+    public String extract(HttpServletRequest request) {
+        try {
+            return Arrays.stream(request.getCookies())
+                .filter(cookie -> cookie.getName().equals("token"))
```
token등의 값은 상수로 관리하는게 좋을 것 같네요. :)

fucct replied:
앗 감사합니다! 이런 실수를 하다니..😂

phs1116 commented on src/main/java/wooteco/subway/web/prehandler/validator/PasswordMatchValidator.java:
```diff
@@ -0,0 +1,31 @@
+
+import wooteco.subway.web.exception.NotMatchPasswordException;
+
+public class PasswordMatchValidator implements ConstraintValidator<PasswordMatch, Object> {
```
패스워드 밸리데이터를 분리하셨네요. :)

phs1116 commented:
- 인수테스트와 api 테스트 둘 다 예외사항에 대한 테스트를 추가해야 하는지, 아니면 하나만 정해서 추가하면 되는지 정확하지 않네요.. 만약 정해서 한다면 어느곳에 두는게 적절할지도 답변주시면 감사하겠습니다

전체 유저스토리가 아닌 api 단위에서의 예외에 대해서는 예외테스트를 작성해주는게 좋을 것 같네요. :)
정확히는 예외 테스트라기 보다는 응답 status code에 대한 테스트라고 볼수도 있겠네요. 

- 인수테스트에서 Junit5의 다이나믹 테스트를 이용하면 유용하다는 걸 알게 되었는데요, 문제는 이전 테스트에서 사용한 변수를 다음 테스트에서 사용할 수 없다는 점이였습니다. 방법이 있는데 제가 찾지 못한걸까요? 아니면 방법이 없지만 감수하고 사용하는 걸까요..?😭

질문에 대해서 이해가 잘 안가는데요, 이전 테스트에서 사용한 변수라는게 어떤 변수를 뜻하는건지 알 수 있을까요? :)


- Spring data Jdbc 관련 질문입니다. 제가 DDD에 대한 기본 지식이 너무 부족해서 어떻게 보면 사소한 질문일수도 있습니다. 또한 미션을 진행하면서 제가 떠오른 궁금증이라 미션과 관련이 깊지 않을 것 같아서 질문 드려도 될 지 모르겠네요..😭 첫번째 질문은 어떠한 엔티티에 진입 가능한 root가 여러개라면, 해당 엔티티는 어떠한 Aggregate에도 포함되지 않는 엔티티인가요? 예를 들면 Article을 통해 그 게시물에 달린 Comment에 접근이 가능하고, User를 통해 User가 작성한 Comment에도 접근 가능하다면, Comment는 독립적인 Aggregate가 되는걸까요..?

이러한 컨텍스트 분리는 관점을 어떻게 가져가냐에 따라 다를 것 같아요. 고민해볼 요소가 많은 부분이구요.
그리고 비즈니스 적인 부분에 대해서 Aggregate를 통한 로직을 풀어내는건 용이하지만,
(비즈니스 관점에서 Aggregate를 풀어나간다면 Member, Article-Comment가 더 적합하겠네요. :)
방금 말씀하신 특정 데이터의 Query 부분에 대해서는 Aggregate로만 한정하기엔 쉽지 않은 부분이 많아요. :)

- 두번째 질문은, 제 질문이 맞다면, 각각의 관계에 있어서 Ref Table이 반드시 존재해야 하나요? 예를 들면 Article 이 삭제될 때 Comment도 같이 삭제되길 원하더라도 독립된 Aggregate이므로 ref Table 을 두고, 로직에서 삭제를 하는게 맞나요? 이렇게되면 하나의 Transaction에서 처리가 불가능할 것 같아서 궁금증이 생겼습니다.

별도의 Aggregate를 분리한다고 해도, Ref Table이 없이 Comment가 직접 Article의 레퍼런스를 가지고 있어도 되는 부분 아닌가요? :)
현재는 같은 서비스 내에 존재하니 트랜잭션에는 큰 문제가 없을 것 같은데, 별도 서비스라면 보상 트랜잭션에 대해서 고민을 해봐야겠죠. :)

phs1116 reviewed (changes requested):
안녕하세요, 디디!
구현 깔끔하게 잘해주셨어요!
피드백 몇가지랑 질문에 대한 답 추가했는데 참고 부탁드릴게요! :)

fucct commented:
> * 인수테스트와 api 테스트 둘 다 예외사항에 대한 테스트를 추가해야 하는지, 아니면 하나만 정해서 추가하면 되는지 정확하지 않네요.. 만약 정해서 한다면 어느곳에 두는게 적절할지도 답변주시면 감사하겠습니다
> 
> 전체 유저스토리가 아닌 api 단위에서의 예외에 대해서는 예외테스트를 작성해주는게 좋을 것 같네요. :)
> 정확히는 예외 테스트라기 보다는 응답 status code에 대한 테스트라고 볼수도 있겠네요.
> 
> * 인수테스트에서 Junit5의 다이나믹 테스트를 이용하면 유용하다는 걸 알게 되었는데요, 문제는 이전 테스트에서 사용한 변수를 다음 테스트에서 사용할 수 없다는 점이였습니다. 방법이 있는데 제가 찾지 못한걸까요? 아니면 방법이 없지만 감수하고 사용하는 걸까요..?😭
> 
> 질문에 대해서 이해가 잘 안가는데요, 이전 테스트에서 사용한 변수라는게 어떤 변수를 뜻하는건지 알 수 있을까요? :)
> 
> * Spring data Jdbc 관련 질문입니다. 제가 DDD에 대한 기본 지식이 너무 부족해서 어떻게 보면 사소한 질문일수도 있습니다. 또한 미션을 진행하면서 제가 떠오른 궁금증이라 미션과 관련이 깊지 않을 것 같아서 질문 드려도 될 지 모르겠네요..😭 첫번째 질문은 어떠한 엔티티에 진입 가능한 root가 여러개라면, 해당 엔티티는 어떠한 Aggregate에도 포함되지 않는 엔티티인가요? 예를 들면 Article을 통해 그 게시물에 달린 Comment에 접근이 가능하고, User를 통해 User가 작성한 Comment에도 접근 가능하다면, Comment는 독립적인 Aggregate가 되는걸까요..?
> 
> 이러한 컨텍스트 분리는 관점을 어떻게 가져가냐에 따라 다를 것 같아요. 고민해볼 요소가 많은 부분이구요.
> 그리고 비즈니스 적인 부분에 대해서 Aggregate를 통한 로직을 풀어내는건 용이하지만,
> (비즈니스 관점에서 Aggregate를 풀어나간다면 Member, Article-Comment가 더 적합하겠네요. :)
> 방금 말씀하신 특정 데이터의 Query 부분에 대해서는 Aggregate로만 한정하기엔 쉽지 않은 부분이 많아요. :)
> 
> * 두번째 질문은, 제 질문이 맞다면, 각각의 관계에 있어서 Ref Table이 반드시 존재해야 하나요? 예를 들면 Article 이 삭제될 때 Comment도 같이 삭제되길 원하더라도 독립된 Aggregate이므로 ref Table 을 두고, 로직에서 삭제를 하는게 맞나요? 이렇게되면 하나의 Transaction에서 처리가 불가능할 것 같아서 궁금증이 생겼습니다.
> 
> 별도의 Aggregate를 분리한다고 해도, Ref Table이 없이 Comment가 직접 Article의 레퍼런스를 가지고 있어도 되는 부분 아닌가요? :)
> 현재는 같은 서비스 내에 존재하니 트랜잭션에는 큰 문제가 없을 것 같은데, 별도 서비스라면 보상 트랜잭션에 대해서 고민을 해봐야겠죠. :)

fucct committed:
refactor: 리뷰 반영 1

- Auth enum 제거
- Interceptor 전체 url 적용
- Favorite equals, hashcode 재정의 삭제
- logger 적용
- transaction 적용
- Favorite을 aggregate root로 수정

fucct commented:
좋은 피드백 감사합니다! 많은걸 배우게 되네요!

두 번째 질문은 다이나믹 테스트의 한 테스트에서 사용한 변수를 다른 테스트에서 쓸 수 있는지에 대한 질문이었습니다! 예를 들자면 첫 번째 테스트에서` int a = 1`를 선언하고 두 번째 테스트에서 `a`변수를 쓸 수 있는 방법이 있는지에 대한 질문이였습니다. 사소한 질문이라 답변 안주셔도 될 것 같습니다!

그리고 마지막 질문에 제가 직접 레퍼런스를 갖는 방법을 놓치고 있어서 질문이 이상했네요 😅 (Member와 Favorite에서 직접 레퍼런스를 갖는 방식으로 수정해보면서 익혔습니다..!) 추가 질문을 드리자면 ref 테이블은 어떤경우에 갖는게 좋을까요..? 답변주시면 감사하겠습니다!

phs1116 commented on src/main/java/wooteco/subway/web/service/favorite/FavoriteService.java:
```diff
@@ -0,0 +1,74 @@
+        favoriteRepository.delete(favorite);
+    }
+
+    private Favorite getFavorite(Member member, FavoriteRequest request) {
```
getFavorite보다는 생성한다는 의미를 가지는게 더 적합하지 않을까요?

fucct replied:
수정했습니다! 자세하게 설명해주셔서 감사합니다!

phs1116 commented:
`추가 질문을 드리자면 ref 테이블은 어떤경우에 갖는게 좋을까요..?`
경우가 다양하겠네요. 양쪽 도메인간의 동일 레벨의 모델이 있을 경우, 서로의 키의 체번이 다르게 될테니 이를 매핑하기 위해 사용할수도 있을 거구요. 하지만 이런 경우도 포함하여 레퍼런스 테이블을 가져가지 않고 모델 내에 레퍼런스 키를 가져는 방식으로 해도 충분한 것 같아요.. :)

phs1116 reviewed (approved):
안녕하세요. 리뷰어 화투입니다.
피드백 반영 잘해주셨어요.
크게 피드백 드릴 내용이 없어서 머지할게요. :)
간단한 피드백이랑 질문 답 추가했는데 참고 부탁드려요!
